### PR TITLE
bpo-32856: Optimize the assignment idiom in comprehensions.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Now `for y in [expr]` in comprehensions is so fast as a simple assignment `y = expr`.

Unlike to the `:=` operator this idiom does not leak a variable to the outer scope.

https://bugs.python.org/issue32856

### CODE REVIEW

File: Doc/whatsnew/3.9.rst [hunk 1]
```diff
@@ -227,6 +227,17 @@ signals to a process using a file descriptor instead of a pid. (:issue:`38712`)
 Optimizations
 =============
 
+* Optimized the idiom for assignment a temporary variable in comprehensions.
+  Now ``for y in [expr]`` in comprehensions is so fast as a simple assignment
```
Comment (scoder) [inline Doc/whatsnew/3.9.rst]: "as fast as" (also in news file)

File: Python/compile.c [hunk 2]
```diff
@@ -4332,12 +4336,38 @@ compiler_sync_comprehension_generator(struct compiler *c,
...
+                break;
+            case Tuple_kind:
+                elts = gen->iter->v.Tuple.elts;
+                break;
+            default:
+                elts = NULL;
+        }
+        if (asdl_seq_LEN(elts) == 1) {
+            expr_ty elt = asdl_seq_GET(elts, 0);
+            if (elt->kind != Starred_kind) {
```
Comment (scoder) [inline Python/compile.c:4355]: Is this condition tested in both cases?

**Review by scoder — COMMENTED** [review]
I don't have an opinion on the general feature, so just a few comments.

File: Python/compile.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Python/compile.c:4355]: Added tests.

File: Doc/whatsnew/3.9.rst (hunk 1 above)
Comment (serhiy-storchaka) [reply Doc/whatsnew/3.9.rst]: Done. Thank you.

**Comment by scoder** [comment]
LGTM

**Pull request merged**